docs/architecture/adr-053-state-sync-prototype.md

```diff
@@ -0,0 +1,198 @@
+
+## Context
+
+State sync will allow a new node to receive a snapshot of the application state without downloading blocks or going through consensus. This bootstraps the node significantly faster than the current fast sync system, which replays all historical blocks.
+
+Background discussions and justifications are detailed in [ADR-042](./adr-042-state-sync.md). Its recommendations can be summarized as:
+
+* The application periodically takes full state snapshots (i.e. eager snapshots).
+
+* The application splits snapshots into smaller chunks that can be individually verified against a chain Merkle root.
+
+* Tendermint uses the light client to obtain a trusted chain Merkle root for verification.
+
+* Tendermint discovers and downloads snapshot chunks in parallel from multiple peers, and passes them to the application via ABCI to be applied and verified against the chain root.
+
+* Historical blocks are not backfilled, so state synced nodes will have a truncated block history.
+
+## Tendermint Proposal
+
+This describes the snapshot/restore process seen from Tendermint. The interface is kept as small and general as possible to give applications maximum flexibility.
+
+### Snapshot Data Structure
+
+A node can have multiple snapshots taken at various heights. Snapshots can be taken in different application-specified formats (e.g. MessagePack as format `1` and Protobuf as format `2`, or similarly for schema versioning). Each snapshot consists of multiple chunks containing the actual state data, allowing parallel downloads and reduced memory usage.
+
+```proto
+message Snapshot {
+    uint64 height = 1;   // The height at which the snapshot was taken
+    uint64 format = 2;   // The application-specific snapshot format
+    uint64 chunks = 3;   // The number of chunks in the snapshot
+    bytes metadata = 4;  // Arbitrary application metadata
+}
+
+message SnapshotChunk {
+    uint64 height = 1;   // The height of the corresponding snapshot
+    uint64 format = 2;   // The application-specific snapshot format
+    uint64 chunk = 3;    // The chunk index (zero-based)
+    bytes data = 4;      // Serialized application state in an arbitrary format
+}
+```
+
+Chunk verification data must be encoded along with the state data in the `data` field.
+
+### ABCI Interface
+
+```proto
+// Lists available snapshots
+message RequestListSnapshots {}
+
+message ResponseListSnapshots {
+    repeated Snapshot snapshots = 1;
+}
+
+// Offers a snapshot to the application
+message RequestOfferSnapshot {
+    Snapshot snapshot = 1;
+}
+
+message ResponseOfferSnapshot {
+    bool accepted = 1;
+}
+
+// Fetches a snapshot chunk
+message RequestGetSnapshotChunk {
+    uint64 height = 1;
+    uint64 format = 2;
+    uint64 chunk = 3;
+}
+
+message ResponseGetSnapshotChunk {
+    SnapshotChunk chunk = 1;
+}
+
+// Applies a snapshot chunk
+message RequestApplySnapshotChunk {
+    SnapshotChunk chunk = 1;
+    bytes chain_hash = 2;
+    bool final = 3;
```

Comment:
why do we need final field if we know the number of chunks?

Reply:
Good point. An earlier proposal didn't have `RequestOfferSnapshot` at all, so the application wouldn't know how many chunks were being restored - `final` is just a vestige from that. Will remove.